### ISSUE TITLE
`unnecessary_cast` quick-fix unexpected result in switch expression

### ISSUE THREAD

**Issue by FMorschel** [dart-lang/linter#4981]
**Describe the issue**
When using the quick-fix for `unnecessary_cast` in a switch expression the removed type cast is the outer one but should be all the inner ones.

**To Reproduce**

```dart
T a<T>(int x, bool test) {
  return switch (x) {
    0 => 0 as T,
    1 => 1 as T,
    _ => throw UnimplementedError()
  } as T;
}
```

**Expected behaviour**

```dart
T a<T>(int x, bool test) {
  return switch (x) {
    0 => 0,    // Removed cast here
    1 => 1,    // Removed cast here
    _ => throw UnimplementedError()
  } as T;
}
```

**Additional context**
This should be taken into account because it is way more verbose to write the cast in every line than at the end.

**Comment by FMorschel** [comment]
Alternatively, the assist to cast a switch statement to a switch expression could move the casts. If the reviewer agrees, I can create a new issue if necessary.

**Comment by srawlins** [comment]
This one's pretty tricky. How things are implemented now, the inner two casts are _not_ unnecessary because `0`'s static type is different from `T` and `1`'s static type is different from `T`. So those casts do something, and they can fail at runtime. Then when we've gotten to the type of the switch expression, it is computed to be `T`, so that outside cast _is_ unnecessary.

But I think broadly your issue raises a valid UX improvement. A list literal is a similar example:

```dart
List<int> f(Object a, Object b) => [a as int, b as int] as List<int>;
```

The outer most cast is considered unnecessary. So we offer a fix to remove it. But we could offer an additional fix to tidy things up internally as well. A fix to "make the cast necessary" by removing inner casts.

Actually, casting a List with a new type argument is dangerous. Maybe better examples would be:

```dart
int g(Object a, Object b) => (1 == 2 ? a as int : b as int) as int;

int h(Object? a, Object b) => (a as int? ?? b as int) as int;
```
